README.md
Python bindings for Turboactivate
=================================

Ctypes-based Python bindings for [LimeLM's](http://wyday.com/limelm/) TurboActivate library.


CHANGELOG.md
# Change Log

All notable changes to this project are documented in this file.

## 1.0.4 - 2016-01-27

### Changed

* Let `TurboActivateConnectionDelayedError` and `TurboActivateConnectionError` bubble up. (Thanks
  to @cbenhagen).


## 1.0.3 - 2014-11-04

### Changed

* `TurboActivate.activate()` throws an exception if activation fails


## 1.0.2 - 2014-11-03

### Changed

* Calling `SetCustomActDataPath` under Linux will raise a runtime exception since the method is not
  available on that platform.


## 1.0.1 - 2014-09-29

### Added

* `TurboActivate.is_date_valid()` now accepts an optional `date` parameter, which must be a
  string.


### Fixed

* We use the correct buffer allocation scheme when using `GetFeatureValue`.'
* Solved a packaging problem (README.rst was not included in the distribution).


## 1.0.0 - 2014-08-12

First public version


example.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#

from turboactivate import (
    TurboActivate,
    GenuineOptions,
    TA_SKIP_OFFLINE,
    TurboActivateError,
    TurboActivateTrialUsedError,
    TurboActivateConnectionError,
    TurboActivateTrialExpiredError,
    TurboActivateTrialCorruptedError,
    TurboActivateConnectionDelayedError
)

# TODO: paste your Version GUID here.
TA_GUID = "18324776654b3946fc44a5f3.49025204"

# TODO: paste the path to your dat file here
TA_DAT = "TurboActivate.dat"

if __name__ == "__main__":
    ta = TurboActivate(TA_DAT, TA_GUID, use_trial=True)
    trial_days = ta.trial_days_remaining()

    print("Trial days remaining %d" % trial_days)

    opts = GenuineOptions()

    # How often to verify with the LimeLM servers (90 days)
    opts.days_between_checks(90)

    # The grace period if TurboActivate couldn't connect to the servers.
    # after the grace period is over IsGenuinEx() will return TA_FAIL instead of
    # TA_E_INET or TA_E_INET_DELAYED
    opts.grace_days(14)

    # In this example we won't show an error if the activation
    # was done offline by passing the TA_SKIP_OFFLINE flag
    opts.flags(TA_SKIP_OFFLINE)

    try:
        ta.is_genuine(opts)
    except (TurboActivateConnectionDelayedError, TurboActivateConnectionError):
        print("YourApp is activated, but it failed to verify the activation with the LimeLM servers. "
              "You can still use the app for the duration of the grace period.")
    except TurboActivateError:
        print("Not activated")

    # TODO: prompt the user for a product key
    try:
        ta.set_product_key("U9MM-4NJ5-QFG8-TWM5-QM75-92YI-NETA")
    except TurboActivateError as e:
        print("key failed to save")

        raise e

    print("Product key saved successfully.")

    ta.activate()

    print ("Activated successfully.")

    # if this app is activated then you can get a feature value (completely optional)
    # See: http://wyday.com/limelm/help/license-features/
    #
    # feature_value = ta.get_feature_value("myFeature")
    # print("the value of myFeaure is %s" % feature_value)


turboactivate/c_wrapper.py
# -*- coding: utf-8 -*-
#
# Copyright 2013, 2014 Develer S.r.l. (http://www.develer.com/)
#
# Author: Lorenzo Villani <lvillani@develer.com>
# Author: Riccardo Ferrazzo <rferrazz@develer.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.

import sys
from os import path as ospath
from ctypes import (
    cdll,
    c_uint,
    c_char_p,
    c_wchar_p,
    Structure,
    create_string_buffer,
    create_unicode_buffer
)

# Utilities

wbuf = create_unicode_buffer if sys.platform == "win32" else create_string_buffer

wstr = c_wchar_p if sys.platform == "win32" else c_char_p

# Wrapper

TA_OK = 0x00000000
TA_FAIL = 0x00000001
TA_E_PKEY = 0x00000002
TA_E_ACTIVATE = 0x00000003
TA_E_INET = 0x00000004
TA_E_INUSE = 0x00000005
TA_E_REVOKED = 0x00000006
TA_E_GUID = 0x00000007
TA_E_PDETS = 0x00000008
TA_E_TRIAL = 0x00000009
TA_E_TRIAL_EUSED = 0x0000000C
TA_E_TRIAL_EEXP = 0x0000000D
TA_E_EXPIRED = 0x0000000D
TA_E_REACTIVATE = 0x0000000A
TA_E_COM = 0x0000000B
TA_E_INSUFFICIENT_BUFFER = 0x0000000E
TA_E_PERMISSION = 0x0000000F
TA_E_INVALID_FLAGS = 0x00000010
TA_E_IN_VM = 0x00000011
TA_E_EDATA_LONG = 0x00000012
TA_E_INVALID_ARGS = 0x00000013
TA_E_KEY_FOR_TURBOFLOAT = 0x00000014
TA_E_INET_DELAYED = 0x00000015
TA_E_FEATURES_CHANGED = 0x00000016
TA_E_ANDROID_NOT_INIT = 0x00000017

# Flags for the UseTrial() and CheckAndSavePKey() functions.

TA_SYSTEM = 0x00000001
TA_USER = 0x00000002

# Flags for the IsGeninueEx() function.

TA_SKIP_OFFLINE = 0x00000001
"""
If the user activated using offline activation
(ActivateRequestToFile(), ActivateFromFile() ), then with this
flag IsGenuineEx() will still try to validate with the LimeLM
servers, however instead of returning TA_E_INET (when within the
grace period) or TA_FAIL (when past the grace period) it will
instead only return TA_OK (if IsActivated()).

If you still want to get the TA_E_INET error code, without
deactivating after the grace period has expired, then use
this flag in tandem with TA_OFFLINE_SHOW_INET_ERR.

If the user activated using online activation then this flag
is ignored.
"""

TA_OFFLINE_SHOW_INET_ERR = 0x00000002
"""
If the user activated using offline activation, and you're
using this flag in tandem with TA_SKIP_OFFLINE, then IsGenuineEx()
will return TA_E_INET on internet failure instead of TA_OK.

If the user activated using online activation then this flag
is ignored.
"""

TA_DISALLOW_VM = 0x00000004
"""
Use the TA_DISALLOW_VM in UseTrial() to disallow trials in virtual machines.
If you use this flag in UseTrial() and the customer's machine is a Virtual
Machine, then UseTrial() will return TA_E_IN_VM.
"""

# Flags for the is_date_valid() Function

TA_HAS_NOT_EXPIRED = 0x00000001


class GENUINE_OPTIONS(Structure):
    _fields_ = [
        ("nLength", c_uint),
        ("flags", c_uint),
        ("nDaysBetweenChecks", c_uint),
        ("nGraceDaysOnInetErr", c_uint),
    ]


class ACTIVATE_OPTIONS(Structure):
    _fields_ = [
        ("nLength", c_uint),
        ("sExtraData", wstr),
    ]


def load_library(path):
    LIBRARIES = {
        'linux2': ospath.join(path, 'libTurboActivate.so'),
        'darwin': ospath.join(path, 'libTurboActivate.dylib'),
        'win32': ospath.join(path, 'TurboActivate.dll'),
    }

    return cdll.LoadLibrary(LIBRARIES[sys.platform])


def validate_result(return_code):
    # All ok, no need to perform error handling.
    if return_code == TA_OK:
        return

    # Raise an exception type appropriate for the kind of error
    if return_code == TA_FAIL:
        raise TurboActivateFailError()
    elif return_code == TA_E_FEATURES_CHANGED:
        raise TurboActivateFeaturesChangedError()
    elif return_code == TA_E_PDETS:
        raise TurboActivateDatFileError()
    elif return_code == TA_E_EDATA_LONG:
        raise TurboActivateExtraDataLongError()
    elif return_code == TA_E_PKEY:
        raise TurboActivateProductKeyError()
    elif return_code == TA_E_INUSE:
        raise TurboActivateInUseError()
    elif return_code == TA_E_REVOKED:
        raise TurboActivateRevokedError()
    elif return_code == TA_E_GUID:
        raise TurboActivateGuidError()
    elif return_code == TA_E_TRIAL:
        raise TurboActivateTrialCorruptedError()
    elif return_code == TA_E_TRIAL_EUSED:
        raise TurboActivateTrialUsedError()
    elif return_code == TA_E_TRIAL_EEXP:
        raise TurboActivateTrialExpiredError()
    elif return_code == TA_E_ACTIVATE:
        raise TurboActivateNotActivatedError()
    elif return_code == TA_E_INVALID_FLAGS:
        raise TurboActivateFlagsError()
    elif return_code == TA_E_COM:
        raise TurboActivateComError()
    elif return_code == TA_E_INET:
        raise TurboActivateConnectionError()
    elif return_code == TA_E_INET_DELAYED:
        raise TurboActivateConnectionDelayedError()
    elif return_code == TA_E_PERMISSION:
        raise TurboActivatePermissionError()

    # Otherwise bail out and raise a generic exception
    raise TurboActivateError()


#
# Exception types
#

class TurboActivateError(Exception):

    """Generic TurboActivate error"""
    pass


class TurboActivateFailError(TurboActivateError):

    """Fail error"""
    pass


class TurboActivateProductKeyError(TurboActivateError):

    """Invalid product key"""
    pass


class TurboActivateNotActivatedError(TurboActivateError):

    """The product needs to be activated."""
    pass


class TurboActivateConnectionError(TurboActivateError):

    """Connection to the server failed."""
    pass


class TurboActivateInUseError(TurboActivateError):

    """The product key has already been activated with the maximum number of computers."""
    pass


class TurboActivateRevokedError(TurboActivateError):

    """The product key has been revoked."""
    pass


class TurboActivateGuidError(TurboActivateError):

    """The version GUID doesn't match that of the product details file."""
    pass


class TurboActivateTrialCorruptedError(TurboActivateError):

    """The trial data has been corrupted, using the oldest date possible."""
    pass


class TurboActivateTrialUsedError(TurboActivateError):

    """The trial extension has already been used."""
    pass


class TurboActivateTrialExpiredError(TurboActivateError):

    """
    The activation has expired or the system time has been tampered
    with. Ensure your time, timezone, and date settings are correct.
    """
    pass


class TurboActivateComError(TurboActivateError):

    """
    The hardware id couldn't be generated due to an error in the COM setup.
    Re-enable Windows Management Instrumentation (WMI) in your group policy
    editor or reset the local group policy to the default values. Contact
    your system admin for more information.

    This error is Windows only.

    This error can also be caused by the user (or another program) disabling
    the "Windows Management Instrumentation" service. Make sure the "Startup type"
    is set to Automatic and then start the service.


    To further debug WMI problems open the "Computer Management" (compmgmt.msc),
    expand the "Services and Applications", right click "WMI Control" click
    "Properties" and view the status of the WMI.
    """
    pass


class TurboActivatePermissionError(TurboActivateError):

    """
    Insufficient system permission. Either start your process as an
    admin / elevated user or call the function again with the
    TA_USER flag instead of the TA_SYSTEM flag.
    """
    pass


class TurboActivateFeaturesChangedError(TurboActivateError):

    """
    If IsGenuine() or IsGenuineEx() reactivated and the features
    have changed, then this will be the return code. Treat this
    as a success.
    """
    pass


class TurboActivateDatFileError(TurboActivateError):

    """The product details file "TurboActivate.dat" failed to load."""
    pass


class TurboActivateFlagsError(TurboActivateError):

    """
    The flags you passed to use_trial(...) were invalid (or missing).
    """
    pass


class TurboActivateExtraDataLongError(TurboActivateError):

    """
    The "extra data" was too long. You're limited to 255 UTF-8 characters.
    Or, on Windows, a Unicode string that will convert into 255 UTF-8
    characters or less.
    """
    pass


class TurboActivateConnectionDelayedError(TurboActivateError):

    """
    is_genuine() previously had a TA_E_INET error, and instead
    of hammering the end-user's network, is_genuine() is waiting
    5 hours before rechecking on the network.
    """
    pass


turboactivate/__init__.py
# -*- coding: utf-8 -*-
#
# Copyright 2013, 2014 Develer S.r.l. (http://www.develer.com/)
#
# Author: Lorenzo Villani <lvillani@develer.com>
# Author: Riccardo Ferrazzo <rferrazz@develer.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.

import sys

from datetime import datetime

from ctypes import pointer, sizeof, c_uint32

from c_wrapper import *

#
# Object oriented interface
#


class GenuineOptions(object):

    """A set of options to use with is_genuine()"""

    FLAG_SKIP_OFFLINE = 0x00000001
    FLAG_OFFLINE_SHOW_INET_ERR = 0x00000002

    def __init__(self, flags=0, grace_days=0, days_between_checks=0):
        self._flags = flags
        self._grace_days = grace_days
        self._days_between_checks = days_between_checks

    def get_pointer(self):
        options = GENUINE_OPTIONS(sizeof(GENUINE_OPTIONS()),
                                  self._flags,
                                  self._days_between_checks,
                                  self._grace_days)
        return pointer(options)

    def flags(self, flags):
        self._flags = flags

    def grace_days(self, days):
        """
        If the call fails because of an internet error,
        how long, in days, should the grace period last (before
        returning deactivating and returning TA_FAIL).

        14 days is recommended.
        """
        self._grace_days = days

    def days_between_checks(self, days):
        """How often to contact the LimeLM servers for validation. 90 days recommended."""
        self._days_between_checks = days


class TurboActivate(object):

    def __init__(self, dat_file, guid, use_trial=False, library_folder="", mode=TA_USER):
        self._lib = load_library(library_folder)
        self._set_restype()

        self.set_current_product(dat_file, guid, use_trial=use_trial, mode=mode)

    #
    # Public
    #

    # Product management

    def set_current_product(self, dat_file, guid, use_trial=False, mode=TA_USER):
        """
        This functions allows you to use licensing for multiple products within
        the same running process.
        """
        self._mode = mode
        self._dat_file = wstr(dat_file)
        self._guid = wstr(guid)

        try:
            self._lib.PDetsFromPath(self._dat_file)
        except TurboActivateFailError:
            # The dat file is already loaded
            pass

        self._lib.SetCurrentProduct(self._guid)

        if use_trial:
            self._lib.UseTrial(self._mode)

    def current_product(self):
        """Gets the "current product" previously set by set_current_product()."""
        buf_size = 128
        buf = wbuf(buf_size)

        self._lib.GetCurrentProduct(buf, buf_size)

        return buf.value

    # Product key

    def product_key(self):
        """
        Gets the stored product key. NOTE: if you want to check if a product key is valid
        simply call is_product_key_valid().
        """
        buf_size = 128
        buf = wbuf(buf_size)

        try:
            self._lib.GetPKey(buf, buf_size)

            return buf.value
        except TurboActivateProductKeyError as e:
            return None

    def set_product_key(self, product_key):
        """Checks and saves the product key."""
        self._lib.CheckAndSavePKey(wstr(product_key), self._mode)

    def blacklists_keys(self, keys_list=[]):
        """
        Blacklists keys so they are no longer valid. Use "BlackListKeys" only if
        you're using the "Serial-only plan" in LimeLM. Otherwise revoke keys.
        """
        arr = (wstr * len(keys_list))()
        arr[:] = keys_list

        self._lib.BlackListKeys(arr, len(arr))

    def is_product_key_valid(self):
        """
        Checks if the product key installed for this product is valid. This does NOT check if
        the product key is activated or genuine. Use is_activated() and is_genuine() instead.
        """
        try:
            self._lib.IsProductKeyValid(self._guid)

            return True
        except TurboActivateError:
            return False

    # Activation status

    def deactivate(self, erase_p_key=True, deactivation_request_file=""):
        """
        Deactivates the product on this computer. Set erase_p_key to True to erase the stored
        product key, False to keep the product key around. If you're using deactivate to let
        a user move between computers it's almost always best to *not* erase the product
        key. This way you can just use activate() when the user wants to reactivate
        instead of forcing the user to re-enter their product key over-and-over again.
        If deactivation_request_file is specified, then it gets the "deactivation request"
        file for offline deactivation.
        """
        e = '1' if erase_p_key else '0'
        fn = self._lib.DeactivationRequestToFile if deactivation_request_file else self._lib.Deactivate
        args = [wstr(deactivation_request_file)] if deactivation_request_file else []

        args.append(e)

        try:
            fn(*args)
        except TurboActivateNotActivatedError:
            return

    def activate(self, extra_data="", activation_request_file=""):
        """
        Activates the product on this computer. You must call set_product_key()
        with a valid product key or have used the TurboActivate wizard sometime
        before calling this function.
        If activation_request_file is specified, then it gets the "activation request"
        file for offline activation.
        """
        if self.is_activated():
            return False

        fn = self._lib.ActivationRequestToFile if activation_request_file else self._lib.Activate
        args = [wstr(activation_request_file)] if activation_request_file else []

        if extra_data:
            fn = self._lib.ActivationRequestToFileEx if activation_request_file else self._lib.ActivateEx
            options = ACTIVATE_OPTIONS(sizeof(ACTIVATE_OPTIONS()),
                                       wstr(extra_data))
            args.append(pointer(options))

        try:
            fn(*args)

            return True
        except TurboActivateError as e:
            if not activation_request_file:
                self.deactivate(True)

            raise e

    def activate_from_file(self, filename):
        """Activate from the "activation response" file for offline activation."""
        self._lib.ActivateFromFile(wstr(filename))

    def get_extra_data(self):
        """Gets the extra data you passed in using activate()"""
        buf_size = 255
        buf = wbuf(buf_size)

        try:
            self._lib.GetExtraData(buf, buf_size)

            return buf.value
        except TurboActivateFailError:
            return ""

    def is_activated(self):
        """ Checks whether the computer has been activated."""
        try:
            self._lib.IsActivated(self._guid)

            return True
        except TurboActivateError:
            return False

    # Features

    def has_feature(self, name):
        return len(self.get_feature_value(name)) > 0

    def get_feature_value(self, name):
        """Gets the value of a feature."""
        buf_size = self._lib.GetFeatureValue(wstr(name), 0, 0)
        buf = wbuf(buf_size)

        self._lib.GetFeatureValue(wstr(name), buf, buf_size)

        return buf.value

    # Genuine

    def is_genuine(self, options=None):
        """
        Checks whether the computer is genuinely activated by verifying with the LimeLM servers.
        If reactivation is needed then it will do this as well.
        Optionally you can pass a GenuineOptions object to specify more details
        """
        fn = self._lib.IsGenuine
        args = [self._guid]

        if options:
            fn = self._lib.IsGenuineEx

            args.append(options.get_pointer())

        try:
            fn(*args)

            return True
        except TurboActivateFeaturesChangedError:
            return True

    # Trial

    def trial_days_remaining(self):
        """
        Get the number of trial days remaining.
        0 days if the trial has expired or has been tampered with
        (1 day means *at most* 1 day, that is it could be 30 seconds)

        You must have instantiated TurboActivate with the use_trial
        flag to use this function
        """
        days = c_uint32(0)

        self._lib.TrialDaysRemaining(self._guid, pointer(days))

        return days.value

    def extend_trial(self, extension_code):
        """Extends the trial using a trial extension created in LimeLM."""
        self._lib.ExtendTrial(wstr(extension_code))

    # Utils

    def is_date_valid(self, date=None):
        """
        Check if the date is valid
        """
        if not date:
            to_check = datetime.utcnow().strftime("%Y-%m-%d %H-%M-%S")
        else:
            to_check = date

        try:
            self._lib.IsDateValid(wstr(to_check), TA_HAS_NOT_EXPIRED)

            return True
        except TurboActivateFlagsError as e:
            raise e
        except TurboActivateError:
            return False

    def set_custom_path(self, path):
        """
        This function allows you to set a custom folder to store the activation
        data files. For normal use we do not recommend you use this function.

        Only use this function if you absolutely must store data into a separate
        folder. For example if your application runs on a USB drive and can't write
        any files to the main disk, then you can use this function to save the activation
        data files to a directory on the USB disk.

        If you are using this function (which we only recommend for very special use-cases)
        then you must call this function on every start of your program at the very top of
        your app before any other functions are called.

        The directory you pass in must already exist. And the process using TurboActivate
        must have permission to create, write, and delete files in that directory.

        On linux it is not available
        """
        if sys.platform.startswith('linux'):
            raise RuntimeError("set_custom_path is not available under linux")

        self._lib.SetCustomActDataPath(wstr(path))

    def set_custom_proxy(self, address):
        """
        Sets the custom proxy to be used by functions that connect to the internet.

        Proxy address in the form: http://username:password@host:port/

        Example 1 (just a host): http://127.0.0.1/
        Example 2 (host and port): http://127.0.0.1:8080/
        Example 3 (all 3): http://user:pass@127.0.0.1:8080/

        If the port is not specified, TurboActivate will default to using port 1080 for proxies.
        """
        self._lib.SetCustomProxy(wstr(address))

    def _set_restype(self):
        self._lib.PDetsFromPath.restype = validate_result
        self._lib.SetCurrentProduct.restype = validate_result
        self._lib.UseTrial.restype = validate_result
        self._lib.GetCurrentProduct.restype = validate_result
        self._lib.GetPKey.restype = validate_result
        self._lib.CheckAndSavePKey.restype = validate_result
        self._lib.BlackListKeys.restype = validate_result
        self._lib.IsProductKeyValid.restype = validate_result
        self._lib.DeactivationRequestToFile.restype = validate_result
        self._lib.Deactivate.restype = validate_result
        self._lib.Activate.restype = validate_result
        self._lib.ActivationRequestToFile.restype = validate_result
        self._lib.ActivationRequestToFileEx.restype = validate_result
        self._lib.ActivateEx.restype = validate_result
        self._lib.ActivateFromFile.restype = validate_result
        self._lib.GetExtraData.restype = validate_result
        self._lib.IsActivated.restype = validate_result
        self._lib.IsGenuine.restype = validate_result
        self._lib.IsGenuineEx.restype = validate_result
        self._lib.TrialDaysRemaining.restype = validate_result
        self._lib.ExtendTrial.restype = validate_result
        self._lib.IsDateValid.restype = validate_result
        self._lib.IsDateValid.restype = validate_result
        self._lib.SetCustomProxy.restype = validate_result

        # SetCustomActDataPath is not defined under linux
        if not sys.platform.startswith('linux'):
            self._lib.SetCustomActDataPath.restype = validate_result
